experiments/cpu-stress/experiment/cpu-stress.go

```diff
@@ -0,0 +1,280 @@
+// limitations under the License.
+
+package experiment
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"os/exec"
+
+	"github.com/litmuschaos/m-agent/internal/m-agent/errorcodes"
+	logger "github.com/litmuschaos/m-agent/internal/m-agent/log"
+	"github.com/litmuschaos/m-agent/internal/m-agent/messages"
+	"github.com/litmuschaos/m-agent/internal/m-agent/upgrader"
+	"github.com/litmuschaos/m-agent/pkg/cpu"
+	"github.com/litmuschaos/m-agent/pkg/probes"
+)
+
+var (
+	stdout bytes.Buffer
+	stderr bytes.Buffer
+	cmd    *exec.Cmd
+)
+
+// CPUStress listens for the client actions and executes them as appropriate
+func CPUStress(w http.ResponseWriter, r *http.Request) {
+
+	// upgrade the connection to a websocket connection
+	upgrader := upgrader.GetConnectionUpgrader()
+
+	clientMessageReadLogger := logger.GetClientMessageReadErrorLogger()
+	steadyStateCheckErrorLogger := logger.GetSteadyStateCheckErrorLogger()
+	executeExperimentErrorLogger := logger.GetExecuteExperimentErrorLogger()
+	commandProbeExecutionErrorLogger := logger.GetCommandProbeExecutionErrorLogger()
+	invalidActionErrorLogger := logger.GetCommandProbeExecutionErrorLogger()
+	chaosAbortErrorLogger := logger.GetChaosAbortErrorLogger()
+	livenessCheckErrorLogger := logger.GetLivenessCheckErrorLogger()
+	closeConnectionErrorLogger := logger.GetCloseConnectionErrorLogger()
+	chaosRevertErrorLogger := logger.GetChaosRevertErrorLogger()
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		fmt.Fprintf(w, "Failed to establish connection with client, err: %v", err)
+		return
+	}
+
+	for {
+
+		action, reqID, payload, err := messages.ListenForClientMessage(conn)
+		if err != nil {
+
+			if err := messages.SendMessageToClient(conn, "ERROR", reqID, errorcodes.GetClientMessageReadErrorPrefix()+err.Error()); err != nil {
+				clientMessageReadLogger.Printf("Error occured while sending error message to client, err: %v", err)
+			}
+
+			if err := conn.Close(); err != nil {
+				clientMessageReadLogger.Printf("Error occured while closing the connection, err: %v", err)
+			}
+
+			return
+		}
+
+		switch action {
+
+		case "CHECK_STEADY_STATE":
+			if err := cpu.CheckStressNG(); err != nil {
+
+				if err := messages.SendMessageToClient(conn, "ERROR", reqID, errorcodes.GetSteadyStateCheckErrorPrefix()+err.Error()); err != nil {
+					steadyStateCheckErrorLogger.Printf("Error occured while sending error message to client, err: %v", err)
+				}
+
+				if err := conn.Close(); err != nil {
+					steadyStateCheckErrorLogger.Printf("Error occured while closing the connection, err: %v", err)
+				}
+
+				return
+			}
+
+			if err := messages.SendMessageToClient(conn, "ACTION_SUCCESSFUL", reqID, nil); err != nil {
+
+				steadyStateCheckErrorLogger.Printf("Error occured while sending feedback message to client, err: %v", err)
+
+				if err := conn.Close(); err != nil {
+					steadyStateCheckErrorLogger.Printf("Error occured while closing the connection, err: %v", err)
+				}
+
+				return
```

Comment:
We can have this as a function so to follow DRY.

Reply:
Can you please point out the part that you think is repetitive here?

Reply:
Correction on my end, we can have this as a function to reduce the number of lines of code and ease up manageability. What I am suggesting is that sending a message to the client and closing the connection are fixed processes in every scenario, if we put them into a function, it will be a one-line addition to every case then. For example
`return handleClientResponse(connection, message, reqid, error)`
and then the function can be what we are doing here
```
func handleClientResponse(...) {
    if err := messages.SendMessageToClient(conn, message, reqID, errorcodes.GetSteadyStateCheckErrorPrefix()+err.Error()); err != nil {
 		steadyStateCheckErrorLogger.Printf("Error occured while sending error message to client, err: %v", err)
 	} // Can put a check for error parameter being nil
	if err := conn.Close(); err != nil {
 			steadyStateCheckErrorLogger.Printf("Error occured while closing the connection, err: %v", err)
 	}
}
```

If this change seems logical and useful, we can take this. If it doesn't fare well with the existing guidelines, we can skip this.

Reply:
Sure, will refactor for this in a separate PR.

Reply:
Let's open issues for anything we use to plan for future!